### DISCUSSION TITLE
オセロの石を置く場所を探す方法

### DISCUSSION THREAD

**Discussion by m2mt-l** [Recursion-Group-G/Othello#47] [Ideas]
とりあえずそんなに最適化は考えていない方法です。その他案があれば提案ください！

# 案1
## 準備
- スタックを2つ用意(myStack、opponentStack)
- SquareにisPlaced: booleanを追加
- placedCounterを設定
- Configにvertical、horizontal、diagonalを追加(探索方向)

## 処理内容概要
- 左上(ここはどこでも可)から3方向(縦、横、斜め)に探索を走らせていく。戻りもこれで実施。
- 完了したら一つ下と一つ右に座標を移動しそれぞれで処理を走らせる(二分木使えるかも)
- 下に移動した座標の処理は縦方向の探索は実施しない
- 右に移動した座標の処理は横方向の処理を実施しない

## 処理内容詳細
方向による違いはないので、以下処理は一つのメソッドの想定(入力で探索方向はコントロール)。
1. 連結リストを探索
2. isEmpty===trueであればスキップ
3. 自分の石(Color.id)と同じのが出ればスタックに入れ始める(myStack)
4. 相手の石は自分の石がスタックにある時点で初めてスタックに入る(opponentStack)
5. 以下のいずれかが現れたら終了
- myStack、opponentStackにスタックした状態でempty=trueが現われれば石を置ける⇒isPlaced=trueにする、placedCounter++(placedCounter) 
- myStackかopponentStackのどちらか片方にしかスタックがない状態でisEmpty===trueが現れたらその行の探索終了
- スタックが空のまま最後まで終了
6. スタックを空にする(本当はキャッシュして石を返す時にうまく使いたい)
7. 逆方向に同じことを実施(スタックが空のまま最後まで終了した場合はスキップ)
8. 全部終了後、placedCounter === 0 だった場合ターンスキップ

**Comment by mcnLeandro** [comment]
これは土曜日にwhimsical使いながら確認したいですね！！

**Reply by mcnLeandro** [reply]
一応自分は大体流れを把握しました！
最適でなくても確実にできそうですね！

ここから新しいアイデアも生まれそうです！

**Reply by m2mt-l** [reply]
考えてみたら、座標移動後は斜め2タイプの方向が必要ですね。。。でも1点で実施する方向は3つで問題ないと思います！これですべて網羅できるかと。

**Comment by mcnLeandro** [comment]
> isPlaced

こちらはisEmptyと違いがなければどちらかにした方がいいかもですね！

**Reply by m2mt-l** [reply]
isEmptyだけで処理できるのであればそれでも問題ないです！ただエラー処理は入れておく必要がある気がします(石があるのにisEmptyがtrueになる等)

**Reply by mcnLeandro** [reply]
そこですよね。。
メソッド作って解決したいですが、モデルは状態維持だけなのでいけなくて、
今はcomputedなどで値を切り替えよう。となっていますが、

これも何かの手違いでStoneが存在するのにisEmpty = true みたいになる可能性を考えずにはいられないですね。

確実にやるなら、回りくどいですが、Helperクラスを作成してそこで判定させるしかなさそうです。
```js
const square = new Square()
Helper.isEmpty(square)
```

**Comment by mcnLeandro** [comment]
playerクラスにスタック持たせますか？？

**Reply by m2mt-l** [reply]
Player毎に持つ必要があると思うので場所はPlayerが適切かと思います！

**Comment by kuro1221** [comment]
自分も考えてみました。

# 案2-①

## 処理内容

1. isEmptyがtrueのSquareを順番に全て、連結リストを使用して8方向探索する
2. Squareが相手の石ならplacedCounter+1し、次のSquareを探索
    - もしSquareがnullなら探索終了し、別の方向の探索を開始する
    - もし次の石が自分の石なら、placedCounterの数と方向をCanPutStoneオブジェクトとして、Squareに記憶させる
        
        ```jsx
        class CanPutStone {
        	top: number;  //返せる石の数(placedCounter)だけ格納
        	topRight: number;
        	~省略~
        	bottomRight: number;
        }
        ```
        
3. 石が置かれたらCanPutStoneの位置情報を元にひっくり返す

## メリット

- 各Square毎にひっくり返せる石の数が記憶されているので、一番ひっくり返せる場所に石を置くというAIが作成しやすい

## デメリット

- isEmptyがtrueのSquareを順番に全て調べるので、序盤はかなり計算量が多い



# 案2-②

案2-①はisEmptyがtrueのSquareを順番に全て調べるが計算量がかなり多いため、盤面を仮想的に縮小して探索する
https://whimsical.com/6YrTUaVgeaZS8tgFPNu8jL
（マス目の黄色の部分のみ探索する）

- 一番左上と一番右下の石が置かれているSquareのPointをBoardに記録する
    - 初期状態では一番左上のSquareのPoint(TopLeftPoint)が(x:3,y:3)、一番右下のSquareのPoint(BottomRightPoint)が(x:4,y:4)
    - for文を使って黄色の部分のマス目のみ8方向探索する
    
    ```jsx
    for (let i = TopLeftPoint.x - 1: i < BottomRightPoint.x + 1; i++) {
    	for (let j = TopLeftPoint.y - 1; j < BottomRighePoint.y + 1; j++){
    		searchSquare = Board.squares[i][j]; //searchSquareを8方向探索
    	}
    }
    ```
    
    - 石が置かれたら左上と右下のPointを更新する

**Reply by m2mt-l** [reply]
案1とのあわせ技でいけそうですね！探索は自分の方法で実施して(3方向)、この案を取り入れれば計算量削減と処理の拡張ができそうです！

**Comment by mcnLeandro** [comment]
notion11/13 まとめ: https://www.notion.so/11-13-256f2d198f6940ea8aa231d68c3ff0e0
whimsical: https://whimsical.com/6YrTUaVgeaZS8tgFPNu8jL
